리펙토링

kimcno3 opened the pull request:
- 제네릭 타입을 통해 하나의 DTO 클래스로 리턴 타입을 지정했다.
- GameController.play()의 경우 게임 종료 여부에 따라 리턴 타입이 달라야 하는 차이점이 있어 상속으로 이를 해결
  - ResponseGameDto <- ResponseGameCloseDto

Commit:
Response DTO 클래스 통합

- 제네릭 타입을 통해 하나의 DTO 클래스로 리턴 타입을 지정했다.
- GameController.play()의 경우 게임 종료 여부에 따라
  리턴 타입이 달라야 하는 차이점이 있어 상속으로 이를 해결
- ResponseGameDto <- ResponseGameCloseDto

f-lab-bright commented on src/main/java/project/baseball/dtos/response/ResponseGameCloseDto.java:
```diff
@@ -1,18 +1,17 @@
-  private boolean success;
-  private GameAnswerDataDto data;
-  private GameAnswerErrorDto error;
+public class ResponseGameCloseDto<T, S> extends ResponseGameDto<T> {
```
ResponseGameDto를 이용해서도 에러 응답을 같이 표현할 수 있지 않을까요??

Commit:
카운트 계산 반복문 메소드로 분리

Commit:
제네릭 타입 선언

f-lab-bright commented:
응답 객체는 어제 같이 리펙토링 하면서 보여드린 방식으로 수정해보시면 좋겠네요 :)

Commit:
리펙토링

- ResponseDto 클래스 하나로 통합
- Dto클래스 내부 클래스 선언
- Controller의 일부 로직 수정으로 Service에 역할 위임

Commit:
리펙토링

- ResponseDto from() 매서드 생성
- GameData build 로직 이동

Commit:
리펙토링

- GameData 메소드로 카운트 계산 메소드 구현
- 정답을 맞춘 경우 다시 한번 정답을 제출했을 때 남은 횟수를 0으로 만들어
  게임 종료 메세지를 리턴

Commit:
리펙토링

- Controller.play() 리턴값을 하나의 메소드로 통합

kimcno3 commented:
정답 완료한 시점엔 게임 결과 Response를 리턴하고, 그 이후로는 게임 종료 Response를 리턴하는 로직을 구현한 방식을 추가 설명하자면
- 정답횟수가 0 이하일 경우에만 게임 종료 응답 메세지 리턴
- 정답을 맞춘 이후에 다시 한번 숫자를 입력했을 때, 이미 정답을 맞춘 기록이 있다면 남은 정답 횟수를 0으로 만들어 종료 응답 메세지가 리턴되도록 구현

그리고 하나의 ResponseDto 클래스에 내부 클래스로 모든 Reponse 객체를 관리하도록 작성했습니다!

Commit:
리펙토링

-  로그코드 삭제
- Service 클래스에서 saveHistory() 메소드 삭제

Commit:
리펙토링

- 예외처리 기능 추가

f-lab-bright commented on src/main/java/project/baseball/repository/GameRepositoryImpl.java:
```diff
@@ -24,23 +22,15 @@ public class GameRepositoryImpl implements GameRepository {
-  public GameData findById(Long id) {
-    return database.get(id);
+  public GameData saveHistory(GameData gameData, GameHistory history) {
+    gameData.getHistories().add(gameData.getHistories().size(), history);
```
getHistories() 에서 add()를 하는 것 보다, gameData.addHistory(history); 같은 형태로 구성되는게 좀 더 좋지 않을까요??
`tell don't ask` 할 수 있도록 해보면 좋겠네요 :)

f-lab-bright commented on src/main/java/project/baseball/dtos/response/ResponseDto.java:
```diff
@@ -0,0 +1,115 @@
+    }
+  }
+
+  /** . */
+
+  @Getter
+  @AllArgsConstructor
+  public static class AnswerData {
+    private boolean correct;
+    private int remainingCount;
+    private int strike;
+    private int ball;
+    private int out;
+
+    private static AnswerData from(GameData gameData) {
+      GameHistory gameHistory = gameData.getHistories().get(gameData.getHistories().size() - 1);
+      GameResult result = gameHistory.getResult();
+
+      boolean correct = gameData.isCorrect();
+      int remainingCount = gameData.getRemainingCount();
+      int s = result.getStrike();
+      int b = result.getBall();
+      int o = result.getOut();
+
+      return new AnswerData(correct, remainingCount, s, b, o);
+    }
+  }
+
+  /** . */
+
+  @Getter
+  @AllArgsConstructor
+  public static class ResultData {
+    private int remainingCount;
+    private int answerCount;
+
+    private static ResultData from(GameData gameData) {
+      return new ResultData(gameData.getRemainingCount(), gameData.getAnswerCount());
+    }
+  }
```
실제 응답에 실릴 data 선언은 파일과 클래스를 분리해서 사용하는게 좀 더 좋을 것 같습니다 :)

Commit:
리펙토링

- ResponseDTO 내부 클래스 파일 분리
- GameData 클래스 메소드 추가 생성(Tell don't ask 원칙)

kimcno3 commented:
주신 피드백 내용 관련해서 수정해봤습니다!

f-lab-bright reviewed (approved):
LGTM with Comments

처음보다 코드가 많이 정리됐네요! :)
객체지향 책에 있던 패턴이나 solid 원칙을 적용해볼 수 있을만한 곳이 있을까요??
